molpipeline/mol2mol/filter.py

```diff
@@ -120,38 +169,212 @@ def pretransform_single(self, value: RDKitMol) -> OptionalMol:
         OptionalMol
             Molecule if it contains only allowed elements, else InvalidInstance.
         """
-        unique_elements = set(atom.GetAtomicNum() for atom in value.GetAtoms())
-        if not unique_elements.issubset(self.allowed_element_numbers):
-            forbidden_elements = unique_elements - self.allowed_element_numbers
+        to_process_value = (
+            Chem.AddHs(value) if 1 in self.allowed_element_numbers else value
+        )
+
+        elements_list = [atom.GetAtomicNum() for atom in to_process_value.GetAtoms()]
+        elements_counter = Counter(elements_list)
+        if any(
+            element not in self.allowed_element_numbers for element in elements_counter
+        ):
             return InvalidInstance(
-                self.uuid,
-                f"Molecule contains following forbidden elements: {forbidden_elements}",
-                self.name,
+                self.uuid, "Molecule contains forbidden chemical element.", self.name
             )
+        for element, (min_count, max_count) in self.allowed_element_numbers.items():
+            count = elements_counter[element]
+            if (min_count is not None and count < min_count) or (
+                max_count is not None and count > max_count
+            ):
+                return InvalidInstance(
+                    self.uuid,
+                    f"Molecule contains forbidden number of element {element}.",
+                    self.name,
+                )
         return value
 
 
-class MixtureFilter(_MolToMolPipelineElement):
-    """MolToMol which removes molecules composed of multiple fragments."""
+class SmartsFilter(_BasePatternsFilter):
+    """Filter to keep or remove molecules based on SMARTS patterns.
+
+    Notes
+    -----
+    There are four possible scenarios:
+        - mode = "any" & keep_matches = True: Needs to match at least one filter element.
+        - mode = "any" & keep_matches = False: Must not match any filter element.
+        - mode = "all" & keep_matches = True: Needs to match all filter elements.
+        - mode = "all" & keep_matches = False: Must not match all filter elements.
+    """
+
+    def _pattern_to_mol(self, pattern: str) -> RDKitMol:
+        """Convert SMARTS pattern to RDKit molecule.
+
+        Parameters
+        ----------
+        pattern: str
+            SMARTS pattern to convert.
+
+        Returns
+        -------
+        RDKitMol
+            RDKit molecule.
+        """
+        return Chem.MolFromSmarts(pattern)
+
 
-    def __int__(
+class SmilesFilter(_BasePatternsFilter):
+    """Filter to keep or remove molecules based on SMILES patterns.
+
```

Comment:
If we keep the SmilesFilter, then comment in the docstring here saying that "In contrast to the SMARTSFilter, which also can match SMILES, the SmilesFilter checks kekulized bonds for aromaticity and then sets it to aromatic while the SmartsFilter detects alternating single and double bonds."